File: app/src/main/java/com/concrete/challenge/githubjavapop/ui/repository/RepositoryViewModel.java
package com.concrete.challenge.githubjavapop.ui.repository;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.concrete.challenge.githubjavapop.api.Api;
import com.concrete.challenge.githubjavapop.api.SingleSchedulers;
import com.concrete.challenge.githubjavapop.domain.RepositoriesResponse;
import com.concrete.challenge.githubjavapop.domain.Repository;

import java.util.ArrayList;

import io.reactivex.disposables.CompositeDisposable;

public class RepositoryViewModel extends ViewModel {
    private MutableLiveData<ArrayList<Repository>> repositories;
    private MutableLiveData<Throwable> error;
    private Api api;
    private SingleSchedulers singleSchedulers;
    private CompositeDisposable disposable;

    public RepositoryViewModel(Api api, SingleSchedulers singleSchedulers) {
        this.api = api;
        this.singleSchedulers = singleSchedulers;
        disposable = new CompositeDisposable();
        repositories = new MutableLiveData<>();
        error = new MutableLiveData<>();
    }

    public void loadRepositories(int page) {
        disposable.add(api.getRepositories(page)
                .compose(singleSchedulers.applySchedulers())
                .subscribe(this::onSuccess, this::onError));
    }

    private <T> void onSuccess(T result) {
        RepositoriesResponse repositoriesResponse = (RepositoriesResponse) result;
        ArrayList<Repository> temp = repositories.getValue();
        if(temp == null) temp = new ArrayList<>();
        temp.addAll(repositoriesResponse.items);
        repositories.setValue(temp);
    }

    private void onError(Throwable error) {
        this.error.setValue(error);
    }

    public LiveData<ArrayList<Repository>> getRepositories() {
        return repositories;
    }

    public LiveData<Throwable> getError() {
        return error;
    }
}

File: app/src/main/java/com/concrete/challenge/presentation/viewmodel/RepositoryViewModel.kt
package com.concrete.challenge.presentation.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.concrete.challenge.domain.io.APIService
import com.concrete.challenge.domain.io.response.RepositoriesResponse
import com.concrete.challenge.utils.callService

class RepositoryViewModel(private val service: APIService) : ViewModel() {

    val repositoriesResponse: MutableLiveData<RepositoriesResponse> by lazy {
        MutableLiveData<RepositoriesResponse>()
    }

    private var page: Int = 1

    fun incrementPage() {
        page++
    }

    fun getRepositories() {
        callService(liveData = repositoriesResponse) { service.getRepositories(page) }
    }

}